kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/CheckpointInstance.java

```diff
@@ -117,14 +179,27 @@ public boolean equals(Object o) {
         if (o == null || getClass() != o.getClass()) {
             return false;
         }
+
+        // For V2 checkpoints, compare the filepath.
         CheckpointInstance checkpointInstance = (CheckpointInstance) o;
+        if (checkpointInstance.format == CheckpointFormat.V2 && format == CheckpointFormat.V2 &&
+                !filePath.equals(checkpointInstance.filePath)) {
+            return false;
+        }
         return version == checkpointInstance.version &&
-                Objects.equals(numParts, checkpointInstance.numParts);
+                Objects.equals(numParts, checkpointInstance.numParts) &&
+                format == checkpointInstance.format;
```

Comment:
Can we use `this.compareTo(o) == 0` instead of comparing individual fields?

Reply:
How would that handle the case that we don't want to compare filepaths for non-V2 checkpoints (since multipart checkpoints are considered part of the same instance)?